### PR TITLE
Wait for current tracer provider to shutdown when dropping telemetry

### PR DESCRIPTION
**Pull request by garypen**
We spawn a detached thread currently. That's probably not a good idea since the router may terminate before the thread terminates and that will likely cause spans to be lost.

Let's wait for the shutdown for up to 5 seconds in the `drop` and then print a warning message if there is an issue.

The figure of 5 seconds is not configurable, but could be made so in future if required.

### CODE REVIEW

File: apollo-router/src/plugins/telemetry/mod.rs [hunk 1]
```diff
@@ -183,9 +183,13 @@ impl Drop for Telemetry {
...
         if count < 2 {
-            std::thread::spawn(|| {
+            // We don't want telemetry to drop until the shutdown completes.
+            // Ignore the result, we log errors as warnings.
+            let _ = std::thread::spawn(|| {
                 opentelemetry::global::shutdown_tracer_provider();
-            });
+            })
+            .join()
+            .map_err(|e| ::tracing::warn!("unable to join tracer shutdown thread: {:?}", e));
```
Comment (bnjjj) [inline apollo-router/src/plugins/telemetry/mod.rs:192]: I'm wondering if we should put a timeout on this ?

Comment (garypen) [reply apollo-router/src/plugins/telemetry/mod.rs:192]: That's a good suggestion. I could add a variation on this: https://stackoverflow.com/questions/36181719/what-is-the-correct-way-in-rust-to-create-a-timeout-for-a-thread-or-a-function
I'll take a look at doing that.
Only thing is knowing how long to wait... (I'll pick 5 seconds for now)

Comment (Geal) [reply apollo-router/src/plugins/telemetry/mod.rs:192]: starting a thread to immediately join on it is the same as calling the code directly. The thread is not necessary (unless you add the timeout)

Comment (garypen) [reply apollo-router/src/plugins/telemetry/mod.rs:192]: I agree. I had a conversation with Bryn about why we even needed the thread at all... Anyway, I'm updating it to add a timeout as per @bnjjj suggestion.

Comment (garypen) [reply apollo-router/src/plugins/telemetry/mod.rs]: done: https://github.com/apollographql/router/pull/2191/commits/7ca2897d3264df7d049d8f8cc1cccccac569924d

**Review by bnjjj — APPROVED** [review]

**Pull request merged**